In [2]:
import pandas as pd
from kg_analytics import *
from IPython.display import display

prime_kg_df = pd.read_csv("c:/diploma/Disease-Gene-Prioritization-Project/primekg/kg.csv", low_memory=False)

In [2]:
calc_global_stats(prime_kg_df)

{'total_nodes': 129375,
 'total_edges': 8100498,
 'num_node_types': 10,
 'num_edge_types': 30,
 'node_types': ['anatomy',
  'biological_process',
  'cellular_component',
  'disease',
  'drug',
  'effect/phenotype',
  'exposure',
  'gene/protein',
  'molecular_function',
  'pathway'],
 'edge_types': ['anatomy_anatomy',
  'anatomy_protein_absent',
  'anatomy_protein_present',
  'bioprocess_bioprocess',
  'bioprocess_protein',
  'cellcomp_cellcomp',
  'cellcomp_protein',
  'contraindication',
  'disease_disease',
  'disease_phenotype_negative',
  'disease_phenotype_positive',
  'disease_protein',
  'drug_drug',
  'drug_effect',
  'drug_protein',
  'exposure_bioprocess',
  'exposure_cellcomp',
  'exposure_disease',
  'exposure_exposure',
  'exposure_molfunc',
  'exposure_protein',
  'indication',
  'molfunc_molfunc',
  'molfunc_protein',
  'off-label use',
  'pathway_pathway',
  'pathway_protein',
  'phenotype_phenotype',
  'phenotype_protein',
  'protein_protein'],
 'density': 0.000483965

In [3]:
connected_components_summary(prime_kg_df)

{'num_components': 1,
 'largest_component_size': 129375,
 'smallest_component_size': 129375,
 'component_sizes': [129375],
 'isolated_node_count': 0}

In [4]:
node_type_stats(prime_kg_df)

,node_type,node_count,mean_degree,median_degree,std_degree,min_degree,max_degree,p90_degree,p99_degree,isolated_nodes
0,biological_process,28642,17.610642,8.0,56.365663,2,3312,32.0,182.00,0
1,gene/protein,27671,190.179538,142.0,213.370458,2,11160,412.0,848.60,0
2,disease,17080,39.958314,12.0,99.283282,2,3048,96.0,386.00,0
3,effect/phenotype,15311,33.583175,6.0,136.054573,2,4798,52.0,551.80,0
4,anatomy,14035,223.178340,2.0,2326.460891,2,34710,6.0,552.56,0
5,molecular_function,11169,17.319903,4.0,254.823517,2,25092,20.4,210.00,0
6,drug,7957,705.214528,26.0,1039.257026,2,5200,2370.0,3904.00,0
7,cellular_component,4176,44.589080,4.0,394.398994,2,10814,37.0,611.50,0
8,pathway,2516,37.930048,18.0,59.801911,2,958,98.0,260.00,0
9,exposure,818,22.826406,8.0,55.044784,2,682,52.0,243.28,0


In [5]:
relation_type_stats(prime_kg_df)

,relation,src_type,dst_type,edge_count,relation_density,unique_heads,unique_tails,avg_tails_per_head,avg_heads_per_tail,relation_class
0,drug_drug,drug,drug,2672628,0.146035,4278,4278,624.737728,624.737728,N-N
1,anatomy_protein_present,gene/protein,anatomy,1518203,0.379791,24829,161,61.146361,9429.832298,N-N
2,anatomy_protein_present,anatomy,gene/protein,1518203,0.379791,161,24829,9429.832298,61.146361,N-N
3,protein_protein,gene/protein,gene/protein,642150,0.001906,18354,18354,34.986924,34.986924,N-N
4,disease_phenotype_positive,effect/phenotype,disease,150317,0.002370,9404,6744,15.984368,22.288998,N-N
5,disease_phenotype_positive,disease,effect/phenotype,150317,0.002370,6744,9404,22.288998,15.984368,N-N
6,bioprocess_protein,gene/protein,biological_process,144805,0.000623,18650,12459,7.764343,11.622522,N-N
7,bioprocess_protein,biological_process,gene/protein,144805,0.000623,12459,18650,11.622522,7.764343,N-N
8,bioprocess_bioprocess,biological_process,biological_process,105772,0.000129,28642,28642,3.692899,3.692899,N-N
9,cellcomp_protein,cellular_component,gene/protein,83402,0.002453,1756,19363,47.495444,4.307287,N-N


In [15]:
deduped_prime_kg_df = prime_kg_df.drop_duplicates(subset=["x_name", "relation", "y_name"])
print(f"Original triples: {len(prime_kg_df)}, Deduped triples: {len(deduped_prime_kg_df)}")

Original triples: 8100498, Deduped triples: 8099991


In [16]:
relation_type_stats(deduped_prime_kg_df)

,relation,src_type,dst_type,edge_count,relation_density,unique_heads,unique_tails,avg_tails_per_head,avg_heads_per_tail,relation_class
0,drug_drug,drug,drug,2672628,0.146035,4278,4278,624.737728,624.737728,N-N
1,anatomy_protein_present,gene/protein,anatomy,1518203,0.379791,24829,161,61.146361,9429.832298,N-N
2,anatomy_protein_present,anatomy,gene/protein,1518203,0.379791,161,24829,9429.832298,61.146361,N-N
3,protein_protein,gene/protein,gene/protein,642150,0.001906,18354,18354,34.986924,34.986924,N-N
4,disease_phenotype_positive,disease,effect/phenotype,150317,0.002370,6744,9404,22.288998,15.984368,N-N
5,disease_phenotype_positive,effect/phenotype,disease,150312,0.002370,9404,6744,15.983837,22.288256,N-N
6,bioprocess_protein,gene/protein,biological_process,144805,0.000623,18650,12459,7.764343,11.622522,N-N
7,bioprocess_protein,biological_process,gene/protein,144805,0.000623,12459,18650,11.622522,7.764343,N-N
8,bioprocess_bioprocess,biological_process,biological_process,105772,0.000129,28642,28642,3.692899,3.692899,N-N
9,cellcomp_protein,cellular_component,gene/protein,83402,0.002453,1756,19363,47.495444,4.307287,N-N


In [11]:
prime_kg_df.shape

(8100498, 12)

In [3]:
from pykeen.triples import TriplesFactory
from pykeen.triples.leakage import unleak

triples_df = prime_kg_df.copy()

labeled_triples = triples_df[["x_name", "relation", "y_name"]].to_numpy(dtype=str)

In [4]:
labeled_triples.shape

(8100498, 3)

In [ ]:
triples_factory = TriplesFactory.from_labeled_triples(
    triples=labeled_triples,
    # create_inverse_triples=False,
    # filter_out_candidate_inverse_relations=True
    # filter_out_candidate_inverse_relations (bool) – Whether to remove triples with relations with the inverse suffix.
)

In [6]:
# filter_out_candidate_inverse_relations=True - nem használ
triples_factory.num_triples, triples_factory.num_entities, triples_factory.num_relations

(8099991, 129262, 30)

In [34]:
training, validation, testing = triples_factory.split(
            ratios=[0.8, 0.1, 0.1],
            random_state=42,
        )

train, validate, test = unleak(
    training,  
    *[validation, testing],
    minimum_frequency=1
)


In [17]:
training.num_triples, validation.num_triples, testing.num_triples

(6479992, 809999, 810000)

In [35]:
train.num_triples, validate.num_triples, test.num_triples

(6479992, 809999, 810000)

In [39]:
from pykeen.datasets import EagerDataset
from pykeen.datasets.analysis import get_relation_count_df

dataset = EagerDataset(training=training, testing=test)
dataset2 = EagerDataset(training=train, testing=test)

before = get_relation_count_df(dataset)
after = get_relation_count_df(dataset2)

merged = before.merge(after, on="relation_id", suffixes=("_before", "_after"))
merged["difference"] = merged["count_before"] - merged["count_after"]

display(merged.sort_values("difference", ascending=False).head(20))

,relation_id,count_before,relation_label_before,count_after,relation_label_after,difference
0,0,26636,anatomy_anatomy,26636,anatomy_anatomy,0
1,1,36441,anatomy_protein_absent,36441,anatomy_protein_absent,0
2,2,2730357,anatomy_protein_present,2730357,anatomy_protein_present,0
3,3,97745,bioprocess_bioprocess,97745,bioprocess_bioprocess,0
4,4,260298,bioprocess_protein,260298,bioprocess_protein,0
5,5,9135,cellcomp_cellcomp,9135,cellcomp_cellcomp,0
6,6,149717,cellcomp_protein,149717,cellcomp_protein,0
7,7,55271,contraindication,55271,contraindication,0
8,8,59522,disease_disease,59522,disease_disease,0
9,9,2175,disease_phenotype_negative,2175,disease_phenotype_negative,0


Na most vagy a split odafigyel hogy ne legyen leak vagy az unleak nem jó

In [5]:
from pykeen.triples.leakage import Sealant  
  
# Identify and remove inverse relations  
sealant = Sealant(triples_factory, minimum_frequency=0.97)  
clean_factory = sealant.apply(triples_factory)

print(f"Relations to delete: {sealant.relations_to_delete}")  
  
# Convert to labels if using TriplesFactory  
if hasattr(triples_factory, 'relation_id_to_label'):  
    deleted_labels = [triples_factory.relation_id_to_id[rel_id] for rel_id in sealant.relations_to_delete]  
    print(f"Deleted relation labels: {deleted_labels}")

Relations to delete: set()
Deleted relation labels: []


In [ ]:
# Before cleaning
print(f"Original: {triples_factory.num_relations} relations, {triples_factory.num_triples} triples")

# After cleaning
print(f"Cleaned: {clean_factory.num_relations} relations, {clean_factory.num_triples} triples")


# df = clean_factory.tensor_to_df(clean_factory.mapped_triples)
# print(df.head())  
# print(f"Unique relations in cleaned data: {df['relation_label'].nunique()}")

Original: 30 relations, 8099991 triples
Cleaned: 30 relations, 8099991 triples


In [ ]:
# 1. Normalize & Hash Set (Fastest, O(n))

# For each triple (h, r, t), create a canonical form: (min(h, t), max(h, t), r)
# Use a set to track seen normalized triples
# Keep only the first occurrence of each canonical form
# Skip when you encounter the normalized pair again

# 2. Bidirectional Lookup Index (O(n) with parallel processing)

# Build a map: {(min(h,t), max(h,t), r) → [list of original triples]}
# For each canonical form with multiple triples, keep one and flag others for deletion
# Good if you need to know which inverse pair corresponds to which triple

# 3. Sort-based Grouping (O(n log n), good for huge KGs)

# Sort all triples by (min(h, t), max(h, t), r) as composite key
# Iterate once and skip consecutive duplicates
# Memory efficient, scales well to very large graphs

# 4. Vectorized NumPy Approach (Fastest for massive datasets)

# Load triples as structured NumPy array with fields: head, tail, relation
# Create canonical column: (np.minimum(h, t), np.maximum(h, t), r)
# Use np.unique() with axis or fancy indexing to deduplicate in one pass
# Can handle millions of triples efficiently

In [7]:
def filter_implicit_inverses(df, h_name="x_name", t_name="y_name", 
                             h_type="x_type", t_type="y_type", rel="relation"):
    """
    Filters out implicit inverse triples by enforcing a deterministic direction.
    Rule 1: Sort by node type alphabetically.
    Rule 2: If types are the same, sort by node name alphabetically.
    """
    
    # Extract as numpy arrays for speed
    hn = df[h_name].astype(str).values
    tn = df[t_name].astype(str).values
    ht = df[h_type].astype(str).values
    tt = df[t_type].astype(str).values
    
    # Boolean mask: True if the current edge is ALREADY in the correct canonical order
    # (head type < tail type) OR (types are equal AND head name <= tail name)
    is_canonical = (ht < tt) | ((ht == tt) & (hn <= tn))
    
    # Construct the canonical head, tail, and types using the mask
    canon_head = np.where(is_canonical, hn, tn)
    canon_tail = np.where(is_canonical, tn, hn)
    canon_htype = np.where(is_canonical, ht, tt)
    canon_ttype = np.where(is_canonical, tt, ht)
    
    # Create a temporary dataframe to run deduplication
    temp = pd.DataFrame({
        'canon_head': canon_head,
        'canon_tail': canon_tail,
        'canon_htype': canon_htype,
        'canon_ttype': canon_ttype,
        'relation': df[rel].values,
        '_original_idx': df.index.values
    })
    
    # Drop duplicates on the canonical representation
    # 'keep="first"' ensures we just keep one of the edges (effectively deleting the reverse)
    dedup_idx = temp.drop_duplicates(
        subset=['canon_head', 'canon_tail', 'relation'], 
        keep='first'
    )['_original_idx'].values
    
    # Return the original dataframe filtered down to the remaining indices
    return df.loc[dedup_idx].reset_index(drop=True)

In [20]:
new_filtered_df = filter_implicit_inverses(deduped_prime_kg_df)
relation_type_stats(new_filtered_df)

,relation,src_type,dst_type,edge_count,relation_density,unique_heads,unique_tails,avg_tails_per_head,avg_heads_per_tail,relation_class
0,anatomy_protein_present,gene/protein,anatomy,1518203,0.379791,24829,161,61.146361,9429.832298,N-N
1,drug_drug,drug,drug,1336314,0.091315,4269,3428,313.027407,389.823221,N-N
2,protein_protein,gene/protein,gene/protein,321075,0.001305,14012,17558,22.914288,18.286536,N-N
3,disease_phenotype_positive,disease,effect/phenotype,149720,0.002444,6558,9342,22.830131,16.026547,N-N
4,bioprocess_protein,gene/protein,biological_process,144805,0.000623,18650,12459,7.764343,11.622522,N-N
5,cellcomp_protein,gene/protein,cellular_component,83402,0.002453,19363,1756,4.307287,47.495444,N-N
6,disease_protein,gene/protein,disease,80411,0.001583,9080,5593,8.855837,14.377078,N-N
7,molfunc_protein,gene/protein,molecular_function,69530,0.000877,18148,4369,3.831276,15.914397,N-N
8,drug_effect,drug,effect/phenotype,64784,0.057554,1137,990,56.978012,65.438384,N-N
9,bioprocess_bioprocess,biological_process,biological_process,52886,0.000136,13619,28641,3.883251,1.846514,N-N


In [17]:
def remove_inverse_relations(
    df: DataFrame,
    head_col: str = "x_name",
    tail_col: str = "y_name",
    relation_col: str = "relation",
) -> DataFrame:
    out = df.copy()
    left = out[head_col].astype(str)
    right = out[tail_col].astype(str)

    canonical_left = left.where(left <= right, right)
    canonical_right = right.where(left <= right, left)

    out = out.assign(_canonical_left=canonical_left, _canonical_right=canonical_right)
    out = out.drop_duplicates(subset=["_canonical_left", "_canonical_right", relation_col])
    return out.drop(columns=["_canonical_left", "_canonical_right"]).reset_index(drop=True)

In [21]:
# === Vectorized set-based dedupe (fast!) ===
def remove_inverse_relations_fast(df, head_col="x_name", tail_col="y_name", relation_col="relation"):
    """
    Vectorized removal of inverse triples.
    Treats (h, r, t) and (t, r, h) as identical by canonical ordering.
    ~1000x faster than iterrows approach.
    """
    # Extract columns as numpy arrays
    heads = df[head_col].astype(str).values
    tails = df[tail_col].astype(str).values
    rels = df[relation_col].astype(str).values
    
    # Vectorized: compute canonical form for all rows at once
    canonical_left = np.where(heads <= tails, heads, tails)
    canonical_right = np.where(heads <= tails, tails, heads)
    
    # Create a temporary dataframe for deduplication
    temp = pd.DataFrame({
        'canonical_left': canonical_left,
        'canonical_right': canonical_right,
        'relation': rels,
        '_idx': range(len(df))
    })
    
    # Keep only first occurrence of each canonical triple
    dedup_idx = temp.drop_duplicates(subset=['canonical_left', 'canonical_right', 'relation'], keep='first')['_idx'].values
    
    return df.iloc[dedup_idx].reset_index(drop=True)

# Test it
print(f"Original shape: {deduped_prime_kg_df.shape}")
import time
start = time.time()
test_set_dedup_fast = remove_inverse_relations_fast(deduped_prime_kg_df)
elapsed = time.time() - start
print(f"After vectorized set-based dedup: {test_set_dedup_fast.shape}")
print(f"Rows removed: {deduped_prime_kg_df.shape[0] - test_set_dedup_fast.shape[0]}")
print(f"Time: {elapsed:.2f}s")

# Compare with dataframe-based approach
start = time.time()
test_df_dedup = remove_inverse_relations(deduped_prime_kg_df)
elapsed_df = time.time() - start
print(f"\nDataframe-based dedup: {test_df_dedup.shape}")
print(f"Time: {elapsed_df:.2f}s")
print(f"Speedup: {elapsed_df / elapsed:.1f}x")
print(f"Same result? {test_set_dedup_fast.shape == test_df_dedup.shape}")

Original shape: (8099991, 12)
After vectorized set-based dedup: (4049998, 12)
Rows removed: 4049993
Time: 3.87s

Dataframe-based dedup: (4049998, 12)
Time: 4.11s
Speedup: 1.1x
Same result? True


No dedup

Original shape: (8100498, 12)  
After vectorized set-based dedup: (4049998, 12)  
Rows removed: 4050500  
Time: 3.87s  

Dataframe-based dedup: (4049998, 12)  
Time: 4.05s  
Speedup: 1.0x  
Same result? True

In [13]:
relation_type_stats(test_set_dedup_fast)

,relation,src_type,dst_type,edge_count,relation_density,unique_heads,unique_tails,avg_tails_per_head,avg_heads_per_tail,relation_class
0,anatomy_protein_present,gene/protein,anatomy,1518203,0.379791,24829,161,61.146361,9429.832298,N-N
1,drug_drug,drug,drug,1336314,0.091315,4269,3428,313.027407,389.823221,N-N
2,protein_protein,gene/protein,gene/protein,321075,0.001305,14012,17558,22.914288,18.286536,N-N
3,disease_phenotype_positive,disease,effect/phenotype,149720,0.002444,6558,9342,22.830131,16.026547,N-N
4,bioprocess_protein,gene/protein,biological_process,144805,0.000623,18650,12459,7.764343,11.622522,N-N
5,cellcomp_protein,gene/protein,cellular_component,83402,0.002453,19363,1756,4.307287,47.495444,N-N
6,disease_protein,gene/protein,disease,80411,0.001583,9080,5593,8.855837,14.377078,N-N
7,molfunc_protein,gene/protein,molecular_function,69530,0.000877,18148,4369,3.831276,15.914397,N-N
8,drug_effect,drug,effect/phenotype,64784,0.057554,1137,990,56.978012,65.438384,N-N
9,bioprocess_bioprocess,biological_process,biological_process,52886,0.000136,13619,28641,3.883251,1.846514,N-N


In [19]:
relation_type_stats(test_df_dedup)

,relation,src_type,dst_type,edge_count,relation_density,unique_heads,unique_tails,avg_tails_per_head,avg_heads_per_tail,relation_class
0,anatomy_protein_present,gene/protein,anatomy,1518203,0.379791,24829,161,61.146361,9429.832298,N-N
1,drug_drug,drug,drug,1336314,0.091315,4269,3428,313.027407,389.823221,N-N
2,protein_protein,gene/protein,gene/protein,321075,0.001305,14012,17558,22.914288,18.286536,N-N
3,disease_phenotype_positive,disease,effect/phenotype,149720,0.002444,6558,9342,22.830131,16.026547,N-N
4,bioprocess_protein,gene/protein,biological_process,144805,0.000623,18650,12459,7.764343,11.622522,N-N
5,cellcomp_protein,gene/protein,cellular_component,83402,0.002453,19363,1756,4.307287,47.495444,N-N
6,disease_protein,gene/protein,disease,80411,0.001583,9080,5593,8.855837,14.377078,N-N
7,molfunc_protein,gene/protein,molecular_function,69530,0.000877,18148,4369,3.831276,15.914397,N-N
8,drug_effect,drug,effect/phenotype,64784,0.057554,1137,990,56.978012,65.438384,N-N
9,bioprocess_bioprocess,biological_process,biological_process,52886,0.000136,13619,28641,3.883251,1.846514,N-N


In [25]:
EXPLICIT_DIRECTIONS = {
    "anatomy_protein_present":    ("anatomy", "gene/protein"),
    "disease_phenotype_positive": ("disease", "effect/phenotype"),
    "bioprocess_protein":         ("biological_process", "gene/protein"),
    "cellcomp_protein":           ("cellular_component", "gene/protein"),
    "disease_protein":            ("disease", "gene/protein"),
    "molfunc_protein":            ("molecular_function", "gene/protein"),
    "drug_effect":                ("drug", "effect/phenotype"),
    "pathway_protein":            ("pathway", "gene/protein"),
    "contraindication":           ("drug", "disease"),
    "drug_protein":               ("drug", "gene/protein"),
    "anatomy_protein_absent":     ("anatomy", "gene/protein"),
    "indication":                 ("drug", "disease"),
    "phenotype_protein":          ("effect/phenotype", "gene/protein"),
    "off-label use":              ("drug", "disease"),
    "exposure_disease":           ("exposure", "disease"),
    "exposure_bioprocess":        ("exposure", "biological_process"),
    "exposure_protein":           ("exposure", "gene/protein"),
    "disease_phenotype_negative": ("disease", "effect/phenotype"),
    "exposure_molfunc":           ("exposure", "molecular_function"),
    "exposure_cellcomp":          ("exposure", "cellular_component"),
}


# DOESENT FLIP THE OTHER VALUES JUST THE NAME!!!
def enforce_explicit_directions(df, target_mapping, h_name="x_name", t_name="y_name", 
                                h_type="x_type", t_type="y_type", rel="relation"):
    df_out = df.copy()
    
    hn = df_out[h_name].astype(str).values
    tn = df_out[t_name].astype(str).values
    ht = df_out[h_type].astype(str).values
    tt = df_out[t_type].astype(str).values
    rels = df_out[rel].astype(str).values
    
    # We will build a boolean mask of rows that are "backwards" and need to be flipped
    swap_mask = np.zeros(len(df_out), dtype=bool)
    
    # 2. Force Explicit Rule Directions
    for relation, (req_head_type, req_tail_type) in target_mapping.items():
        rel_mask = (rels == relation)
        # If the row is this relation, but the head type matches the required TAIL type, it's backwards!
        needs_swap = rel_mask & (ht == req_tail_type) & (tt == req_head_type)
        swap_mask = swap_mask | needs_swap

    # 3. Handle Same-Type Relations (e.g., drug_drug, disease_disease)
    # For these, we just enforce alphabetical order of the node names so they can be deduplicated.
    same_type_mask = (ht == tt)
    alpha_swap = same_type_mask & (hn > tn)
    swap_mask = swap_mask | alpha_swap
    
    # Optional 4: Fallback for entirely unlisted relations connecting different types
    # (Alphabetical by node type just to guarantee deduplication works)
    unlisted_mask = ~(same_type_mask | np.isin(rels, list(target_mapping.keys())))
    fallback_swap = unlisted_mask & (ht > tt)
    swap_mask = swap_mask | fallback_swap
    
    # 5. Actively OVERWRITE the dataframe arrays with the flipped values
    df_out[h_name] = np.where(swap_mask, tn, hn)
    df_out[t_name] = np.where(swap_mask, hn, tn)
    df_out[h_type] = np.where(swap_mask, tt, ht)
    df_out[t_type] = np.where(swap_mask, ht, tt)
    
    # 6. Now that every inverse has been physically flipped to match the exact same string...
    # drop_duplicates will perfectly swallow the inverses and keep exactly 1 copy.
    df_out = df_out.drop_duplicates(subset=[h_name, rel, t_name]).reset_index(drop=True)
    
    return df_out

# Execute:
final_kg = enforce_explicit_directions(prime_kg_df, target_mapping=EXPLICIT_DIRECTIONS)
relation_type_stats(final_kg)

,relation,src_type,dst_type,edge_count,relation_density,unique_heads,unique_tails,avg_tails_per_head,avg_heads_per_tail,relation_class
0,anatomy_protein_present,anatomy,gene/protein,1518203,0.379791,24829,161,61.146361,9429.832298,N-N
1,drug_drug,drug,drug,1336314,0.091315,4269,3428,313.027407,389.823221,N-N
2,protein_protein,gene/protein,gene/protein,321075,0.001305,14012,17558,22.914288,18.286536,N-N
3,disease_phenotype_positive,disease,effect/phenotype,150317,0.002187,6993,9827,21.495352,15.296326,N-N
4,bioprocess_protein,biological_process,gene/protein,144805,0.000623,18650,12459,7.764343,11.622522,N-N
5,cellcomp_protein,cellular_component,gene/protein,83402,0.002453,19363,1756,4.307287,47.495444,N-N
6,disease_protein,disease,gene/protein,80411,0.001583,9080,5593,8.855837,14.377078,N-N
7,molfunc_protein,molecular_function,gene/protein,69530,0.000877,18148,4369,3.831276,15.914397,N-N
8,drug_effect,drug,effect/phenotype,64784,0.057554,1137,990,56.978012,65.438384,N-N
9,bioprocess_bioprocess,biological_process,biological_process,52886,0.000136,13619,28641,3.883251,1.846514,N-N


In [ ]:
def enforce_explicit_directions_safe(df, target_mapping, rel="relation"):
    df_out = df.copy()
    
    # We use these specific columns just to determine IF we need to swap
    ht = df_out["x_type"].astype(str).values
    tt = df_out["y_type"].astype(str).values
    hn = df_out["x_name"].astype(str).values
    tn = df_out["y_name"].astype(str).values
    rels = df_out[rel].astype(str).values
    
    # Boolean mask tracking which rows must be entirely flipped
    swap_mask = np.zeros(len(df_out), dtype=bool)
    
    # 1. Flip based on Explicit Rule Directions
    for relation, (req_head_type, req_tail_type) in target_mapping.items():
        rel_mask = (rels == relation)
        needs_swap = rel_mask & (ht == req_tail_type) & (tt == req_head_type)
        swap_mask = swap_mask | needs_swap

    # 2. Same-Type Relations (e.g., drug_drug): Sort alphabetically by name
    same_type_mask = (ht == tt)
    alpha_swap = same_type_mask & (hn > tn)
    swap_mask = swap_mask | alpha_swap
    
    # 3. Fallback for unlisted relations connecting different types: Sort alphabetically by type
    unlisted_mask = ~(same_type_mask | np.isin(rels, list(target_mapping.keys())))
    fallback_swap = unlisted_mask & (ht > tt)
    swap_mask = swap_mask | fallback_swap
    
    # 4. SAFELY SWAP THE ENTIRE NODE PAYLOAD
    # Find all x_ columns that have a matching y_ column
    x_cols = [c for c in df_out.columns if c.startswith("x_")]
    valid_pairs = [(x_col, "y_" + x_col[2:]) for x_col in x_cols if "y_" + x_col[2:] in df_out.columns]
    
    for x_col, y_col in valid_pairs:
        # Extract the original payload columns
        x_vals = df_out[x_col].values.copy()
        y_vals = df_out[y_col].values.copy()
        
        # Overwrite the dataframe: if swap_mask is True, put Y in X, otherwise keep X
        df_out[x_col] = np.where(swap_mask, y_vals, x_vals)
        df_out[y_col] = np.where(swap_mask, x_vals, y_vals)
    
    # 5. Finally, deduplicate using the fully canonicalized names
    # Subset must include all columns we care about deduplicating on
    df_out = df_out.drop_duplicates(subset=["x_name", rel, "y_name"]).reset_index(drop=True)
    
    return df_out

# Execute safely:
final_kg = enforce_explicit_directions_safe(prime_kg_df, target_mapping=EXPLICIT_DIRECTIONS)
relation_type_stats(final_kg)

In [32]:
final_kg.shape

(4049998, 12)

In [27]:
final_kg.head()

,relation,display_relation,x_index,x_id,x_type,x_name,x_source,y_index,y_id,y_type,y_name,y_source
0,protein_protein,ppi,0,9796,gene/protein,KIF15,NCBI,8889,56992,gene/protein,PHYHIP,NCBI
1,protein_protein,ppi,1,7918,gene/protein,GPANK1,NCBI,2798,9240,gene/protein,PNMA1,NCBI
2,protein_protein,ppi,2,8233,gene/protein,TTC33,NCBI,5646,23548,gene/protein,ZRSR2,NCBI
3,protein_protein,ppi,3,4899,gene/protein,MAN1B1,NCBI,11592,11253,gene/protein,NRF1,NCBI
4,protein_protein,ppi,4,5297,gene/protein,PI4KA,NCBI,2122,8601,gene/protein,RGS20,NCBI


In [33]:
final_kg.to_csv("c:/diploma/Disease-Gene-Prioritization-Project/primekg/kg_filtered.csv", index=False)

In [ ]:
# === Verify: show examples of deduped pairs ===
# SUPER SLOW
# Find pairs where (h, r, t) and (t, r, h) both exist
def find_inverse_pairs(df, head_col="x_name", tail_col="y_name", relation_col="relation"):
    """Find all inverse pairs in the dataframe."""
    triples = df[[head_col, relation_col, tail_col]].drop_duplicates()
    pairs_found = []
    
    for h, r, t in triples.values:
        # Look for the reverse
        reverse_exists = ((triples[head_col] == t) & 
                         (triples[relation_col] == r) & 
                         (triples[tail_col] == h)).any()
        if reverse_exists and (t, r, h, "->", h, r, t) not in pairs_found:
            pairs_found.append((h, r, t, "↔", t, r, h))
            print(f"Found inverse pair: {h} --{r}--> {t} ↔ {t} --{r}--> {h}")
    
    return pairs_found

inverse_pairs = find_inverse_pairs(test_df_dedup)
print(f"Total inverse pairs found: {len(inverse_pairs)}")
print(f"\nFirst 10 inverse pairs (h, r, t ↔ t, r, h):")
for i, pair in enumerate(inverse_pairs[:10]):
    h, r, t, arrow, t2, r2, h2 = pair
    print(f"  {i+1}. {h} --{r}--> {t} {arrow} {t2} --{r2}--> {h2}")

print(f"\nThese pairs will be deduplicated to keep only one direction.")

In [9]:
prime_kg_filtered_df = remove_inverse_relations(prime_kg_df)
prime_kg_filtered_df = prime_kg_filtered_df.reset_index(drop=True)
prime_kg_filtered_df.shape

(4050064, 12)

In [12]:
relation_type_stats(prime_kg_filtered_df)

,relation,src_type,dst_type,edge_count,relation_density,unique_heads,unique_tails,avg_tails_per_head,avg_heads_per_tail,relation_class
0,anatomy_protein_present,gene/protein,anatomy,1518203,0.379791,24829,161,61.146361,9429.832298,N-N
1,drug_drug,drug,drug,1336314,0.091315,4269,3428,313.027407,389.823221,N-N
2,protein_protein,gene/protein,gene/protein,321075,0.001305,14012,17558,22.914288,18.286536,N-N
3,disease_phenotype_positive,disease,effect/phenotype,149720,0.002444,6558,9342,22.830131,16.026547,N-N
4,bioprocess_protein,gene/protein,biological_process,144805,0.000623,18650,12459,7.764343,11.622522,N-N
5,cellcomp_protein,gene/protein,cellular_component,83402,0.002453,19363,1756,4.307287,47.495444,N-N
6,disease_protein,gene/protein,disease,80411,0.001583,9080,5593,8.855837,14.377078,N-N
7,molfunc_protein,gene/protein,molecular_function,69530,0.000877,18148,4369,3.831276,15.914397,N-N
8,drug_effect,drug,effect/phenotype,64784,0.057554,1137,990,56.978012,65.438384,N-N
9,bioprocess_bioprocess,biological_process,biological_process,52886,0.000136,13619,28641,3.883251,1.846514,N-N


In [14]:
# === Disease distribution and stratification options ===
# Analyze disease rarity and grouping strategies

disease_df = prime_kg_filtered_df[prime_kg_filtered_df["x_type"] == "disease"]
disease_counts = disease_df.groupby("x_name").size().reset_index(name="edge_count")
disease_counts = disease_counts.sort_values("edge_count", ascending=False).reset_index(drop=True)

print("Disease distribution (by edge count):")
print(f"Total diseases: {len(disease_counts)}")
print(f"Mean edges per disease: {disease_counts['edge_count'].mean():.1f}")
print(f"Median edges per disease: {disease_counts['edge_count'].median():.1f}")
print(f"Min: {disease_counts['edge_count'].min()}, Max: {disease_counts['edge_count'].max()}")
print("\nTop 10 diseases by edge count:")
print(disease_counts.head(10))

# Rarity bins: group diseases by frequency quartiles
disease_counts["rarity_group"] = pd.qcut(disease_counts["edge_count"], q=4, labels=["rare", "uncommon", "common", "very_common"], duplicates="drop")
print("\nDisease rarity groups (quartiles):")
print(disease_counts["rarity_group"].value_counts().sort_index())

# Option 1: Stratify by rarity bin
print("\n=== Stratification Options ===")
print("1. By rarity (quartiles):")
print("   - Ensures each split has rare, common, and hub diseases")
print("   - Good for balanced representation across frequency ranges")

# Option 2: Stratify by edge count rank
disease_counts["rank_group"] = pd.cut(disease_counts.index, bins=5, labels=["top20%", "top40%", "top60%", "top80%", "bottom20%"])
print("\n2. By rank (top vs. tail):")
print("   - Ensures each split has hubs and rare diseases")
print("   - Prevents bias towards hub diseases in any split")

# Option 3: Check if disease type/category is available
if "x_source" in prime_kg_filtered_df.columns:
    print("\n3. By disease source/category (if available):")
    print(disease_df["x_source"].unique()[:10])

print("\nTo stratify splits:")
print("  1. Group diseases by chosen strategy (rarity, rank, or source)")
print("  2. For each disease group, allocate edges: 70% train, 15% val, 15% test")
print("  3. Keep reverse edge pairs together in same split")

Disease distribution (by edge count):
Total diseases: 11013
Mean edges per disease: 16.6
Median edges per disease: 7.0
Min: 1, Max: 1523

Top 10 diseases by edge count:
                                              x_name  edge_count
0                                  Mendelian disease        1523
1                                  syndromic disease         621
2                   X-linked intellectual disability         513
3                 congenital nervous system disorder         474
4                      hereditary spastic paraplegia         460
5                  syndromic intellectual disability         436
6         developmental and epileptic encephalopathy         408
7                        Charcot-Marie-Tooth disease         363
8                    genetic nervous system disorder         356
9  multiple congenital anomalies/dysmorphic syndr...         347

Disease rarity groups (quartiles):
rarity_group
rare           3432
uncommon       2121
common         2766
very_co

ok tehát az lesz hogy először megnézem hogy lehetne még máshogy filterezni, pl 3asok settel ugyan az a 3as mindegy milyen sorrendben ne szerepeljen kétszer

kellene egy jó leakage teszt hasonló a szűréshez?

kipróbálni a pykken unleak()et ha jó akkor az mehet és maradhatnak az iversek, n= ? mit csinál az a param?

teszt a disease groupok stratifiolásához - lehet külön nem foglalkozunk azok stratifiolásával és használjuk tovább a pykkent de azt lecsekoljuk utána hogy a ritka meg gyakori betegségek is ugyan anyira maradtak meg.

a disease grouping még nagy kérdés mert típus szerint akkor kell külső forrás hogy melyik betegség milyen típusú, ritkaság szerint is lehet az lenne a jó megoldás de közben meg vehetjük az is ritkának minek kevés kapcsolata van a gráfban


a végére kb nem is kell sokat változtatni a dolgokon lehet csak az unleak()-et kell beszerelni de ezeket a lépéseket meg gondolatmeneteket mind le lehet írni mint mérnöki megoldás

In [2]:
import pandas as pd

kg_filtered_df = pd.read_csv("c:/diploma/Disease-Gene-Prioritization-Project/primekg/kg_filtered.csv", low_memory=False)
print(f"Original shape: {kg_filtered_df.shape}")
kg_filtered_df = kg_filtered_df.drop_duplicates(subset=["x_name", "relation", "y_name"]).reset_index(drop=True)
print(f"After deduplication: {kg_filtered_df.shape}")

Original shape: (4049998, 12)
After deduplication: (4049998, 12)


In [14]:
from pykeen.triples import TriplesFactory
from pykeen.datasets import Dataset, EagerDataset

labeled_triples = kg_filtered_df[["x_name", "relation", "y_name"]].to_numpy(dtype=str)
all_triples = TriplesFactory.from_labeled_triples(triples=labeled_triples, create_inverse_triples=True)

splits = [0.98, 0.01, 0.01]

training, validation, testing = all_triples.split(
    ratios=splits,
    random_state=42,
)

In [15]:
training.num_triples, validation.num_triples, testing.num_triples

(3968998, 40500, 40500)

In [19]:
training.create_inverse_triples, validation.create_inverse_triples, testing.create_inverse_triples

(True, True, False)

In [18]:
validation.create_inverse_triples = True
validation.create_inverse_triples

True

In [20]:
ds = EagerDataset(training=training, validation=validation, testing=testing)
ds.training.num_triples, ds.validation.num_triples, ds.testing.num_triples

(3968998, 40500, 40500)

In [21]:
ds.training.create_inverse_triples, ds.validation.create_inverse_triples, ds.testing.create_inverse_triples

(True, True, False)